SYCL/ESIMD/api/functional/functions/functions_atomic_update_inc_dec.cpp

```diff
@@ -0,0 +1,239 @@
+
+      cgh.parallel_for<
+          Kernel<DataT, NumElems, TestCaseT, ChosenOperatorT, NDRangeDimT,
+                 NumElemsToChangeT, ChangeElemFilterT, AlgorithmToChangeT>>(
+          nd_range, [=](sycl::nd_item<1> nd_item) SYCL_ESIMD_KERNEL {
+            const size_t work_item_index = nd_item.get_global_linear_id();
+            shared_result_ptr[work_item_index] =
+                TestCaseT::template call_esimd_function<
+                    DataT, NumElemsToChange, ChangeElemFilterT, ChosenOperator>(
+                    shared_init_values_ptr, offsets_ptr);
+          });
+    });
+    queue.wait_and_throw();
+
+    ChangeElemFilterT filter{};
+    size_t num_changed_elems = 0;
+    // Collect the number of elements that should be changed. This value will be
+    // used to calculate expected value for all elements that will be returned
+    // after atomic_update call the sum of changed elements.
+    for (size_t i = 0; i < NumElemsToChange; ++i) {
+      num_changed_elems += (filter(i) == 1);
+    }
+
+    std::sort(shared_result.begin(), shared_result.end());
+    for (size_t i = 0; i < NumberIterations; ++i) {
+      const DataT &retrieved = shared_result[i];
+      const DataT &expected = base_value * num_changed_elems +
+                              (i - NumberIterations) * num_changed_elems;
+      if (expected != retrieved) {
+        passed = fail_test(i, expected, retrieved, data_type,
+                           "atomic_update return value", operator_type);
+      }
+    }
+
+    std::vector<size_t> changed_elems_indexes;
+    // Collect the indexess that has been changed.
+    for (size_t i = 0; i < NumElemsToChange; ++i) {
+      if (filter(i) == 1) {
+        changed_elems_indexes.push_back(local_offsets[i]);
+      }
+    }
+
+    // Push the largest value to avoid the following error: can't dereference
+    // out of range vector iterator.
+    changed_elems_indexes.push_back(std::numeric_limits<size_t>::max());
+    auto updated_elem_next_index = changed_elems_indexes.begin();
+
+    const DataT &expected = base_value;
+    DataT expected_after_change =
+        get_expected_value<ChosenOperator>(base_value, NumberIterations);
+    // Verify that values, that do not was changed has initial values.
+    for (size_t i = 0; i < NumElems; ++i) {
+      // If current index is less than changed element index verify that this
+      // element hasn't been changed.
+      const DataT &retrieved = shared_init_values[i];
+      if (i < *updated_elem_next_index) {
+        if (expected != retrieved) {
+          passed = fail_test(i, expected, retrieved, data_type,
+                             "value that should not be updated", operator_type);
+        }
+      } else {
+        if (expected_after_change != retrieved) {
+          passed = fail_test(i, expected_after_change, retrieved, data_type,
+                             "value that should be updated", operator_type);
+        }
+        updated_elem_next_index++;
```

Comment:
Nit, to avoid a temporary creation
```suggestion
        ++updated_elem_next_index;
```

Reply:
this comment has been applied in this commit: https://github.com/intel/llvm-test-suite/pull/939/commits/6a710603c232774f40423443be7d11b5be6f5013